packages/protocol/contracts/common/UniswapFeeHandlerSeller.sol

```diff
@@ -0,0 +1,158 @@
+import "../stability/interfaces/ISortedOracles.sol";
+import "./UsingRegistry.sol";
+import "../stability/StableToken.sol";
+import "../common/FixidityLib.sol";
+import "openzeppelin-solidity/contracts/token/ERC20/IERC20.sol";
+import "../common/Initializable.sol";
+
+import "../uniswap/interfaces/IUniswapV2RouterMin.sol";
+import "../uniswap/interfaces/IUniswapV2FactoryMin.sol";
+
+contract UniswapFeeHandlerSeller is IFeeHandlerSeller, UsingRegistry, Initializable {
+  using FixidityLib for FixidityLib.Fraction;
+  mapping(address => address[]) public routerAddresses;
+
+  uint256 constant MAX_TIMESTAMP_BLOCK_EXCHANGE = 20;
+
+  event ReceivedQuote(address router, uint256 quote);
+  event RouterUsed(address router);
+
+  /**
+   * @notice Sets initialized == true on implementation contracts.
+   * @param test Set to true to skip implementation initialisation.
+   */
+  constructor(bool test) public Initializable(test) {}
+
+  function initialize(address _registryAddress) external initializer {
+    _transferOwnership(msg.sender);
+    setRegistry(_registryAddress);
+  }
+
+  // This function explicitly defines few variables because it was getting error "stack too deep"
+  function sell(
+    address sellTokenAddress,
+    address buyToken,
+    uint256 amount,
+    uint256 maxSlippage // as fraction,
+  ) external {
+    require(
+      buyToken == registry.getAddressForOrDie(GOLD_TOKEN_REGISTRY_ID),
+      "Buy token can only be gold token"
+    );
+
+    // An improvement to this function would be to allow the user to pass a path as argument
+    // and if it generates a better outcome that the ones enabled that gets used
+    // and the user gets a reward
+
+    IERC20 goldToken = getGoldToken();
+    // address celoAddress = address(getGoldToken());
+
+    uint256 bestRouterIndex = 0;
+    uint256 bestRouterQuote = 0;
+
+    address[] memory path = new address[](2);
+    // address[] memory thisTokenRouterAddresses = routerAddresses[sellTokenAddress];
+
+    require(routerAddresses[sellTokenAddress].length > 0, "routerAddresses should be non empty");
+
+    // IERC20 token = IERC20(sellTokenAddress);
+    uint256 balanceToBurn = IERC20(sellTokenAddress).balanceOf(address(this));
+
+    for (uint256 i = 0; i < routerAddresses[sellTokenAddress].length; i++) {
+      address poolAddress = routerAddresses[sellTokenAddress][i];
+      IUniswapV2RouterMin router = IUniswapV2RouterMin(poolAddress);
+
+      path[0] = sellTokenAddress;
+      path[1] = address(goldToken);
+
+      // using the second return value becuase it's the last argument
+      // the previous values show how many tokens are exchanged in each path
+      // so the first value would be equivalent to balanceToBurn
+      uint256 wouldGet = router.getAmountsOut(balanceToBurn, path)[1];
+      emit ReceivedQuote(poolAddress, wouldGet);
+      if (wouldGet > bestRouterQuote) {
+        bestRouterQuote = wouldGet;
+        bestRouterIndex = i;
```

Comment:
Instead of saving bestRouterIndex we could save 

bestRouterAddress
bestRouter

In such case we wouldn't have to retrieve it from storage on line 89 again

Reply:
great point, this piece of code is not currently in use, so will fix after merging